sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryParallelReader.java

```diff
@@ -0,0 +1,124 @@
+import com.google.cloud.bigquery.v3.ParallelRead.Session;
+import com.google.cloud.bigquery.v3.RowOuterClass.Row;
+import java.io.IOException;
+import java.util.Iterator;
+import org.apache.beam.sdk.io.BoundedSource;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Iterates over all rows assigned to a particular reader in a read session.
+ */
+class BigQueryParallelReader<T> extends BoundedSource.BoundedReader<T> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(BigQueryParallelReader.class);
+
+  private Session session;
+  private final BigQueryServices client;
+  private Iterator<ReadRowsResponse> responseIterator;
+  private Iterator<Row> rowIterator;
+  private final ReadRowsRequest request;
+  private SerializableFunction<SchemaAndRowProto, T> parseFn;
+  private BoundedSource<T> source;
+  private BigQueryOptions options;
+  private Row currentRow;
+
+  BigQueryParallelReader(
+      Session session,
+      ReadLocation readLocation,
+      Integer rowBatchSize,
+      SerializableFunction<SchemaAndRowProto, T> parseFn,
+      BoundedSource<T> source,
+      BigQueryServices client,
+      BigQueryOptions options) {
+    this.session = checkNotNull(session, "session");
+    this.client = checkNotNull(client, "client");
+    this.parseFn = checkNotNull(parseFn, "parseFn");
+    this.source = checkNotNull(source, "source");
+    this.options = options;
+
+    ReadRowsRequest.Builder builder = ReadRowsRequest.newBuilder().setReadLocation(readLocation);
+    if (rowBatchSize != null) {
+      builder.setOptions(ReadOptions.newBuilder().setMaxRows(rowBatchSize).build());
+    }
+
+    this.request = builder.build();
+  }
+
+  /**
+   * Empty operation, the table is already open for read.
+   * @throws IOException on failure
+   */
+  @Override
+  public boolean start() throws IOException {
+    responseIterator = client.getTableReadService(options).readRows(request);
+    return advance();
+  }
+
+  @Override
+  public boolean advance() throws IOException {
+    if (responseIterator == null) {
+      throw new IOException("start() needs to be called first");
+    }
+    if (rowIterator != null && rowIterator.hasNext()) {
+      currentRow = rowIterator.next();
+      return true;
+    }
+    if (rowIterator == null || !rowIterator.hasNext()) {
+      if (responseIterator.hasNext()) {
+        rowIterator = responseIterator.next().getRowsList().iterator();
+        if (rowIterator.hasNext()) {
+          currentRow = rowIterator.next();
+          return true;
+        }
+      }
+    }
+    currentRow = null;
+    return false;
+  }
+
+  @Override
+  public T getCurrent() {
+    if (currentRow == null) {
+      return null;
+    }
+    return parseFn.apply(new SchemaAndRowProto(session.getProjectedSchema(), currentRow));
+  }
+
+  @Override
+  public void close() {
+    // Do nothing.
```

Comment:
Dataflow liquid sharding is not currently supported ? Note that Dataflow may not be able to properly autoscale reading due to this (we'll be limited to the initial set of splits).

Reply:
Dataflow liquid sharding is not currently supported -- you're dependent on our liquid sharding for now. We will support stream splitting for Alpha AFAIK, but not for EAP.

Reply:
We created the max number of sources possible in the initial split. Without the support for sub file sharding and split stream, further split is not possible right now.